### PR TITLE
Initial implementation of general eigenvalue decomposition

### PR DESCRIPTION
**Pull request by Randl**

### CODE REVIEW

**Comment by mrry** [comment]
I'm going to punt this one to @rmlarsen, since linear algebra is more his domain!

File: tensorflow/compiler/tests/BUILD [hunk 1]
```diff
@@ -276,6 +276,24 @@ tf_xla_py_test(
     ],
 )
 
+tf_xla_py_test(
+    name = "eig_op_test",
+    size = "medium",
+    srcs = ["eig_op_test.py"],
```
Comment (mrry) [inline tensorflow/compiler/tests/BUILD]: Does this file exist? I can't see it in the PR or in the repository....

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by Randl** [comment]
I've not got any response from you or anyone else, neither any review expect note about non-existing file, which was fixed. Are there any actions expected from me at this point?

**Comment by Randl** [comment]
@rthadur Also I have no rights to reopen the PR

**Comment by rthadur** [comment]
reopened , can you please resolve conflicts

**Comment by Randl** [comment]
@rthadur done, rebased on master

**Comment by Randl** [comment]
ping

**Comment by mrry** [comment]
I’m sorry you’ve experienced a delay on this pull request. Unfortunately, this code is outside my domain of expertise. As I mentioned on May 2nd, @rmlarsen or someone else more familiar with Eigen would be a more appropriate reviewer.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by Randl** [comment]
ping @rmlarsen 
Can you please review or propose someone else who can review?

**Comment by rthadur** [comment]
This PR will be reviewed soon , thanks for your patience.

**Comment by rmlarsen** [comment]
@Randl my apologies for the long delay. I will review this.

File: tensorflow/python/ops/linalg_ops.py [hunk 2]
```diff
@@ -305,6 +305,49 @@ def _composite_impl(matrix, rhs, l2_regularizer):
...
         matrix, rhs, l2_regularizer, fast=fast, name=name)
 
+@tf_export('linalg.eig', v1=[])
+def eig(tensor, name=None):
+  """Computes the eigen decomposition of a batch of matrices.
+
+  Computes the eigenvalues and eigenvectors of the innermost N-by-N matrices
+  in `tensor` such that
+  `tensor[...,:,:] * v[..., :,i] = e[..., i] * v[...,:,i]`, for i=0...N-1.
+
```
Comment (rmlarsen) [inline tensorflow/python/ops/linalg_ops.py]: We need to mention that the eigenvalues and eigenvectors for a non-Hermitian matrix in general are complex, and the eigenvectors are not guaranteed to be linearly independent (The matrix may be "defective", i.e. have lower geometric than algebraic multiplicity). Furthermore, mention that what we compute here are the _right_ eigenvectors. Maybe add a reference to R. A. Horn & C. R. Johnson, "Matrix Analysis", Chapter 1.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
Overall very nice. Thank you for the contribution, and please accept my apologies for missing this PR.

File: tensorflow/core/ops/linalg_ops.cc [hunk 3]
```diff
@@ -383,6 +383,14 @@ REGISTER_OP("SelfAdjointEig")
       return Status::OK();
     });
 
+REGISTER_OP("Eig")
+    .Input("input: T")
```
[the remark this answers is not in the document]
Comment (Randl) [reply tensorflow/core/ops/linalg_ops.cc]: @rmlarsen I've decomposed the op to input and output types, however it seems that underlying `LinearAlgebraOp` doesn't support different input and output types (at least, straightforward attempt to do so didn't work). The solution I thought about is to decompose `LinearAlgebraOp` similarly to what I did in `ComplexEigOp`, with default of same type for input and output. What do you think?

File: tensorflow/python/kernel_tests/self_adjoint_eig_op_test.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-"""Tests for tensorflow.ops.math_ops.matrix_inverse."""
+"""Tests for tensorflow.ops.linalg_ops.self_adjoint_eig."""
```
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/self_adjoint_eig_op_test.py]: Thanks for the cleanup!

File: tensorflow/core/kernels/eig_op_complex64.cc [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/kernels/eig_op_impl.h"
+
+namespace tensorflow {
+
+REGISTER_LINALG_OP("Eig", (ComplexEigOp<complex64, complex64>),
```
Comment (rmlarsen) [inline tensorflow/core/kernels/eig_op_complex64.cc]: Now that this supports real type as well, should we simply call the kernel "EigOp" for consistency? EigenDecompositionOp would perhaps be less idiomatic, but more informative. I'm fine with either.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Review by rmlarsen — DISMISSED** [review]
Thank you for the nice contribution!

**Comment by rthadur** [comment]
@Randl can you please check the build failures ?

**Comment by Randl** [comment]
@rthadur fixed lint stuff, should work now

**Comment by Randl** [comment]
Sorry, I messed up a bit with version control. The current version worked locally for Ubuntu 18.04.

**Comment by Randl** [comment]
@rmlarsen can you rerun CI again please

**Comment by Randl** [comment]
@rmlarsen Can you do it once again please? Sorry for bothering you, I'm new to the whole API thing. Hopefully, it should be the last one.

**Comment by rmlarsen** [comment]
@Randl no problem. It is rather tedious to add a new op, sorry.

**Comment by Randl** [comment]
@rthadur @rmlarsen Is it ok now? Seems like windows fails are due to upstream. Is anything else required from my side?

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@Randl thank you. We'll take it from here.

**Pull request merged**